mediatek: filogic: wax220: support factory image

agners opened the pull request:
Enable building a factory image which can be flashed through the OEM
firmware's web interface. It seems that the web interface requires a
minimum file size of 10MiB, otherwise it will not accept the image.

The update image is a regular sysupgrade tarball packed in a Netgear
encrypted image. The Netgear encrypted image is the same as used in
WAX202 or WAX206, including the encryption keys and IV.

agners committed:
mediatek: filogic: wax220: support factory image

Enable building a factory image which can be flashed through the OEM
firmware's web interface. It seems that the web interface requires a
minimum file size of 10MiB, otherwise it will not accept the image.

The update image is a regular sysupgrade tarball packed in a Netgear
encrypted image. The Netgear encrypted image is the same as used in
WAX202 or WAX206, including the encryption keys and IV.

This adds a script which creates the rootfs_data volume on first
startup. This is required since the OEM firmware's sysupgrade scripts
do not create such a paritition. Note that any script ordered after
70_initramfs_test will not get executed on initramfs. Hence this new
script 75_rootfs_prepare won't create the rootfs_data volume when
using the recovery initramfs.

Also, this deletes the kernel_backup and rootfs_backup volumes in case
we have to create the rootfs_data volumes. This makes sure that
OpenWrt is the actual backup firmware instead of the stock firmware.

References in WAX220 GPL source:
https://www.downloads.netgear.com/files/GPL/WAX220-V1.0.2.8-gpl-src.tar.gz

* package/base-files/files/lib/upgrade/nand.sh:186
  Creation of rootfs_data is disabled

* Uboot-upstream/board/mediatek/common/ubi_helper.c
  Automatic creation of UBI backup volumes

Signed-off-by: Stefan Agner <stefan@agner.ch>

agners commented:
/cc @Flole998

mrkiko commented:
Why the u-boot envtools config change?

agners commented:
> Why the u-boot envtools config change?

The order of the MTD partitions changed, which changes the mtd number.

agners commented:
> Is it broken in general for this device or is this related to your changes?

It is related to the change (reordering of the mtd partitions causes the U-Boot env to be on a different device).

agners commented:
> Is the u-boot environment partition change also needed when you install it like it was done without your patch? Is it broken in general for this device or is this related to your changes?

Whoops this change got squashed into the wrong commit. I'll fit it, thanks for bringing this up.

agners commented:
Tested on a WAX220 with FW 1.0.3.0: Update to OpenWrt using the factory image works.

The initramfs doesn't seem to require special padding, so dropping the board specific build script.

agners commented:
While OpenWrt starts successfully, I realized that the `rootfs_data` UBI volume is missing when updating via OEM firmware.

```sh
```

agners commented:
> Not sure how to fix this, maybe a custom boot script can be added to create a `rootfs_data` UBI volume if missing on first boot?

I've added `target/linux/mediatek/filogic/base-files/lib/preinit/75_rootfs_prepare` to address this issue. I hope this is an acceptable work around.

Flole998 commented:
This code will also run if openwrt is started as initramfs, right? That's not a good idea, the initramfs should leave the onboard flash memory untouched unless the user explicitly modifies it (by using sysupgrade command for example).

agners commented:
> This code will also run if openwrt is started as initramfs, right?

As per my testing, it appears not. But I don't know why :see_no_evil: . The script is there (/lib/preinit/75_rootfs_prepare), and I remove the volume, so it really should create it. But I don't see a message in the boot log, and the volume isn't being created in my initramfs. I am not exactly sure why it works. Note I took inspiration from `target/linux/bcm4908/base-files/lib/preinit/75_rootfs_prepare`, which also does not have a protection. 

To be sure, we could add this:

```sh
[ "$INITRAMFS" = "1" ] || boot_hook_add preinit_main rootfs_prepare
```

Flole998 commented:
I'd suggest figuring out why it's not running on an initramfs to avoid adding unnecessary code.

agners commented:
> I'd suggest figuring out why it's not running on an initramfs to avoid adding unnecessary code.

Yeah, I am trying, but so far didn't made any progress. A early "echo" in `rootfs_prepare` just doesn't get called in the initrmafs case. Any hints how to debug this?

agners commented:
Ok, I think I understand now: `package/base-files/files/lib/preinit/70_initramfs_test` tests if this is a initramfs and breaks the processing of the hooks. So essentially everything hook registered in 70+ isn't executed. So I think we can rely on this, do you agree?

Flole998 commented:
Yes, it seems to be designed exactly for this. Although this is more difficult to understand, it seems like it's exactly what's wanted in this case.

agners commented:
> Yes, it seems to be designed exactly for this. Although this is more difficult to understand, it seems like it's exactly what's wanted in this case.

I've removed the `INITRAMFS` test again and added a note in the commit message.

Thanks for the review! :pray:

agners commented:
I've also tested this also with `CONFIG_HAS_TESTING_KERNEL=y` set, along with #13017. I've installed the factory image also on a new unit, things seem to work as expected: The image gets flashed to `kernel` and `rootfs` volume, and the `rootfs_data` volume gets created on first boot.

I've added initial documentation for this platform, see https://openwrt.org/toh/netgear/wax220.

@dangowrt, since you reviewed/merged the initial WAX220 support: Could you maybe have a look if this looks fine to you? There is also the cleanup PR #13017.

blocktrron commented:
Merged to my staging tree - thanks!

agners commented:
@blocktrron thanks for merging! Do you have an estimate when this commit will appear on main (and therefor make it into snapshot releases?)

blocktrron commented:
Done.
